File: Complementar - Estruturas de dados (Didático)/Algoritmos08/Alg08.java
/*
♦ Algoritmo 08: Estrutura de Dados Linear Estática (não Dinâmica) de Classificação/Ordenação de Divisão e Conquista Merge Sort (Ordenação pelo Método (Recursivo) de Mistura/Mesclagem).
*/

public class Alg08 {
    public static void mergeSort(int[] array) {
        if (array.length <= 1) {
            return;
        }
        
        int meio = array.length / 2;
        int[] metadeEsquerda = new int[meio];
        int[] metadeDireita = new int[array.length - meio];
        
        // Preencher as duas metades
        for (int i = 0; i < meio; i++) {
            metadeEsquerda[i] = array[i];
        }
        for (int i = meio; i < array.length; i++) {
            metadeDireita[i - meio] = array[i];
        }
        
        // Recursivamente ordenar as duas metades
        mergeSort(metadeEsquerda);
        mergeSort(metadeDireita);
        
        // Mesclar as metades ordenadas
        merge(metadeEsquerda, metadeDireita, array);
    }
    
    private static void merge(int[] metadeEsquerda, int[] metadeDireita, int[] array) {
        int i = 0; // Índice para a metade esquerda
        int j = 0; // Índice para a metade direita
        int k = 0; // Índice para o array final
        
        while (i < metadeEsquerda.length && j < metadeDireita.length) {
            if (metadeEsquerda[i] <= metadeDireita[j]) {
                array[k] = metadeEsquerda[i];
                i++;
            } else {
                array[k] = metadeDireita[j];
                j++;
            }
            k++;
        }
        
        // Copiar os elementos restantes da metade esquerda
        while (i < metadeEsquerda.length) {
            array[k] = metadeEsquerda[i];
            i++;
            k++;
        }
        
        // Copiar os elementos restantes da metade direita
        while (j < metadeDireita.length) {
            array[k] = metadeDireita[j];
            j++;
            k++;
        }
    }
    
    public static void main(String[] args) {
        int[] array = {64, 34, 25, 12, 22, 11, 90};
        
        mergeSort(array);
        
        System.out.println("Array ordenado:");
        for (int i = 0; i < array.length; i++) {
            System.out.print(array[i] + " ");
        }
    }
}

File: Complementar - Estruturas de dados (Didático)/Algoritmos08/Alg08.cpp
/* 
♦ Algoritmo 08: Estrutura de Dados Linear Estática (não Dinâmica) de Classificação/Ordenação de Divisão e Conquista Merge Sort (Ordenação pelo Método (Recursivo) de Mistura/Mesclagem).
*/

#include <iostream>
using namespace std;

void merge(int array[], int metadeEsquerda[], int tamanhoEsquerda, int metadeDireita[], int tamanhoDireita) {
    int i = 0; // Índice para a metade esquerda
    int j = 0; // Índice para a metade direita
    int k = 0; // Índice para o array final
    
    while (i < tamanhoEsquerda && j < tamanhoDireita) {
        if (metadeEsquerda[i] <= metadeDireita[j]) {
            array[k] = metadeEsquerda[i];
            i++;
        } else {
            array[k] = metadeDireita[j];
            j++;
        }
        k++;
    }
    
    // Copiar os elementos restantes da metade esquerda
    while (i < tamanhoEsquerda) {
        array[k] = metadeEsquerda[i];
        i++;
        k++;
    }
    
    // Copiar os elementos restantes da metade direita
    while (j < tamanhoDireita) {
        array[k] = metadeDireita[j];
        j++;
        k++;
    }
}

void mergeSort(int array[], int tamanho) {
    if (tamanho <= 1) {
        return;
    }
    
    int meio = tamanho / 2;
    int tamanhoEsquerda = meio;
    int tamanhoDireita = tamanho - meio;
    int metadeEsquerda[tamanhoEsquerda];
    int metadeDireita[tamanhoDireita];
    
    // Preencher as duas metades
    for (int i = 0; i < tamanhoEsquerda; i++) {
        metadeEsquerda[i] = array[i];
    }
    for (int i = meio; i < tamanho; i++) {
        metadeDireita[i - meio] = array[i];
    }
    
    // Recursivamente ordenar as duas metades
    mergeSort(metadeEsquerda, tamanhoEsquerda);
    mergeSort(metadeDireita, tamanhoDireita);
    
    // Mesclar as metades ordenadas
    merge(array, metadeEsquerda, tamanhoEsquerda, metadeDireita, tamanhoDireita);
}

int main() {
    int array[] = {64, 34, 25, 12, 22, 11, 90};
    int tamanho = sizeof(array) / sizeof(array[0]);
    
    mergeSort(array, tamanho);
    
    cout << "Array ordenado:" << endl;
    for (int i = 0; i < tamanho; i++) {
        cout << array[i] << " ";
    }
    
    return 0;
}
